Verifying tortoise initial algorithm

y0sher opened the pull request:
## Motivation

We want to change block structure to be more compact to reduce mesh size and the size of messages on the wire. 
we want to implement a verifying tortoise to reduce the complexity of the tortoise algo by verifying against the hare result (input vector) when possible and the full algorithm only when needed.

as stated in the [SMIP-0008](https://github.com/spacemeshos/SMIPS/issues/2)

## Changes
wrote initial algorithm with a test.

## Test Plan
- [x] - happy flow test
- [x] - multiple scenarios with different opinion blocks

## TODO
<!-- This section should be removed when all items are complete -->
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

y0sher committed:
add verifying_tortoise.go and change block structure. WIP

y0sher committed:
wip added test and some more code..
brought back normal votes fot testing to work

y0sher committed:
initial tortoise algorithm code

y0sher committed:
remove not used table and change receiver name

y0sher committed:
fix bad baseblock bug. some log changes.

y0sher committed:
fix bug in good block marking. add tests.

y0sher committed:
add more tests and fix some bugs.

y0sher committed:
implemented working eviction. added a few tests.

y0sher committed:
introduced recovery using old mechanism.

y0sher committed:
use tortoise baseblock instead of hare votes

y0sher committed:
tortoise algorithm and tests. add diffs inside the baseblock function by passing the input vector in.

y0sher committed:
add syncing on inputvector and validate layers using it.

y0sher committed:
replace tortoise with verifying tortoise

y0sher committed:
fix func sig

antonlerner commented on mesh/meshdb.go:
```diff
@@ -267,6 +275,40 @@ func (m *DB) SaveContextualValidity(id types.BlockID, valid bool) error {
+		return err
+	}
+
+	return m.inputVector.Put(lyrid.Bytes(), i)
```
bit confusing name for a buffer

y0sher replied:
what  do you mean. it's a database not quite a buffer. I can add DB in the end but that seems redundant. it can also be named hareResults since this is what we save eventually.

antonlerner replied:
the `i` variable which holds the bytes

y0sher committed:
initially set the var to types.LayerID

y0sher committed:
no need for casting here.

y0sher committed:
turbohare now save results on tick

y0sher committed:
formatting and some signature fixes, plus some logs

y0sher committed:
fit tortoise to new genesis. fix test bugs.

antonlerner commented on tortoise/verifying_algorithm.go:
```diff
@@ -0,0 +1,92 @@
+package tortoise
```
some comments wouldn't hurt

y0sher replied:
added

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+
+	bdp blockDataProvider
+
+	Last  types.LayerID
```
should all members be exported?

y0sher replied:
The exporting is needed because currently the recovery mechanism is to save all state to disk represented as XDR and it doesn't include un-exported fields.

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+	t.Verified = genesisLayer.Index()
+}
+
+func (t *turtle) evict() {
```
add comment to what evict does

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+				continue
+			}
+			delete(t.GoodBlocksIndex, id)
+			t.GoodBlocksArr = removeBlockIndex(t.GoodBlocksArr, goodidx)
```
explain what `GoodBlocksArr `

y0sher replied:
changed a bit. added comments.

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+		bdp:                 bdp,
+		Last:                0,
+		avgLayerSize:        avgLayerSize,
+		GoodBlocksArr:       make([]types.BlockID, 0, 10),
```
comment on what all these slices mean

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+//	return support
+//}
+
+func (t *turtle) singleInputVectorFromDB(lyrid types.LayerID, blockid types.BlockID) (vec, error) {
```
rename to `getSingleInputVectorFromDB`

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+	return lyrResult
+}
+
+func (t *turtle) BaseBlock(getres func(id types.LayerID) ([]types.BlockID, error)) (types.BlockID, [][]types.BlockID, error) {
```
should this be exported?

y0sher replied:
yes it is used by the block builder to point on a baseblock.

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+		return lyrResult
+	}
+
+	input := *inputvector
```
is the dereferencing necessary?

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+		for exneu := range b.NeutralDiff {
+			exblk, err := t.bdp.GetBlock(b.NeutralDiff[exneu])
+			if err != nil {
+				panic(fmt.Sprint("err , ", err))
```
write something meaningful in panic

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+		for exag := range b.AgainstDiff {
+			exblk, err := t.bdp.GetBlock(b.AgainstDiff[exag])
+			if err != nil {
+				panic(fmt.Sprint("err , ", err))
```
write something meaningful in panic

antonlerner commented on turbohare/hare.go:
```diff
@@ -5,31 +5,63 @@ import (
+				continue
+			}
 
+			time.Sleep(time.Second * 3)
```
is the sleep necessary?

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+
+		blks, err := t.bdp.LayerBlockIds(i)
+		if err != nil {
+			continue // Panic? can't get layer
```
at least print error

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+		} else {
+			raw, err := t.bdp.GetLayerInputVector(i)
+			if err != nil {
+				// this sets the input to abstain
```
log?

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+				panic(fmt.Sprintf(" database err or layer not exist. %v", i))
+			}
+
+			res, err := getres(i)
```
get what res?

antonlerner replied:
`getRes`

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,602 @@
+
+func (t *turtle) opinionMatches(layerid types.LayerID, opinion2 opinion, getres func(id types.LayerID) ([]types.BlockID, error)) ([]map[types.BlockID]struct{}, error) {
+	// using maps makes it easy to not add duplicates
+	a := make(map[types.BlockID]struct{})
```
a= against? n = neutral? f = ? maybe rename them better

y0sher committed:
fix miner tests. new flow

y0sher committed:
Added comments. moved genesis check upwards so once is enough.

y0sher committed:
remove goodblocks array, we can just use a map. added more comments.

y0sher committed:
add comments and fix lint issues.

y0sher committed:
set evict and last

y0sher committed:
give more time to sync in apptest

y0sher committed:
fix some logs. try to wait actual delta of creating blocks instead of arbitrary time.

y0sher committed:
fix merge conflicts

y0sher committed:
try to fix gomod tidy

y0sher committed:
support all genesis. add debug logs to trtl.

y0sher committed:
set debug to the correct test.

y0sher committed:
count window from new effective genesis.

y0sher committed:
only close the socket from the higher level. if encountered read/write… (#2141)

… error it's already closed.

## Motivation

Fix the bug that occurred in testnet. 
bug flow: once a connection had to be removed because our cache was full. we would call Close on this connection. which will close the closeChan channel. this will result in an error returned inside `MsgConnection`'s reading flow which will call the same `Close` again in result in close of a closed channel.

## Changes

only close the udp socket from the higher level. if encountered read/write error it's already closed.

## Test Plan
added testing.



Co-authored-by: Lane Rettig <lanerettig@gmail.com>

y0sher committed:
added more logs to help understanding what happens in travis.

y0sher committed:
log fixes. pass hare as input vector backup when using turbohare

y0sher committed:
try to fix gomod

y0sher commented:
bors try

antonlerner commented on sync/syncer.go:
```diff
@@ -448,7 +452,11 @@ func (s *Syncer) handleNotSynced(currentSyncLayer types.LayerID) {
 
-		s.ValidateLayer(lyr) // wait for layer validation
+		hareForLayer, err := s.DB.GetLayerInputVector(lyr.Index())
+		if err != nil {
```
isn't hareForlayer already nil when there is an error? what should happen now?

antonlerner commented on tortoise/verifying_tortoise.go:
```diff
@@ -0,0 +1,570 @@
+	return 1
+}
+
+//func (t *turtle) saveOpinion() error {
```
remove?

y0sher committed:
try running with latest sync generated db

y0sher committed:
fix config and file names for new app tests.

y0sher commented:
bors try

y0sher committed:
add the layer field when logging about block building

y0sher committed:
set exception list max size to be hdist*avglayersize

y0sher committed:
fix for forBlockInView to use diffs

y0sher committed:
save contextual validity only when a layer is verified

y0sher commented:
fixed mining system tests. 
TODO: 
[ ] - Address CR
[ ] - Remove traces of old block scheme and tortoise
[ ] - last clean up to verifying tortoise: remove some old unused code.

y0sher commented:
bors try

y0sher committed:
fix mesh tests

y0sher committed:
rename buffer for proper naming

y0sher committed:
fix activation tests

y0sher committed:
fix tortoise tests

y0sher committed:
fixed verifying tortoise tests

y0sher committed:
Export variables for persistence to work

y0sher committed:
fix lint

y0sher commented:
bors try

y0sher commented:
bors try

y0sher committed:
cr and clean up.

y0sher commented:
bors try

y0sher commented:
bors try

y0sher commented:
moved work to #2201
